README.md
# ethbian

The easy way to run a full Ethereum node on Raspberry Pi 4  
**Ethbian = Raspbian + Geth + Grafana**  
Lastest version: **v0.6**  
Homepage: [https://ethbian.org](https://ethbian.org)

## quick start

### installation

#### a) download ethbian image

Already prepared sd-card image can be found [here](https://ethbian.org/downloads/ethbian-v0.6-2020-02-16.img.gz)

| ethbian image file             | sha 256 checksum                                                 |
| ------------------------------ | ---------------------------------------------------------------- |
| ethbian-v0.6-2020-02-16.img.gz | 7f8fd9d7003cc7d80eabf95dd3a934b8be9b86df164e30652d9b196bc1d62ae9 |

To verify the checksum of a downloaded file:

| system | command                  |
| ------ | ------------------------ |
| Mac    | shasum -a 256 _filename_ |
| Linux  | sha256sum _filename_     |

#### b) or create your own image

- download & install Raspbian Buster Lite
- login (user: pi/pass: raspberry) and execute:

> wget https://raw.githubusercontent.com/ethbian/ethbian/master/ethbian_install.sh && bash ethbian_install.sh

- restart the box
- login again (user: pi/pass: ethbian)
- register (free) for a MaxMind account and obtain a license key in order to download GeoLite2 database (free)  
  Visit [MaxMind page](https://www.maxmind.com/en/geolite2/signup) to register, [download](https://www.maxmind.com/en/accounts/214108/geoip/downloads) the GeoLite2City db and  
  save the file as /usr/local/lib/collectd/geolite_city.mmdb

## first steps

default user: **pi**  
defualt pass: **ethbian**

#### what you should know

- swap is disabled
- wifi is disabled
- IPv6 is disabled
- bluetooth is disabled
- sound card is disabled
- camera modules are disabled
- unnecessary services are disabled

#### network configuration

You can use a simple **ethbian-net.sh** script to configure network.  
You can also use standard **raspi-config** tool to do the same.  
Remember that wifi is disabled - use wired connection instead:  
lower latency, reliable connection, speed.

#### SSD disk configuration

You can use the **ethbian-ssd-init.sh** script to initialize a new disk drive  
or remount your partition after upgrading Ethbian.  
If you prefer to do this manually - SSD should be mounted to the **/mnt/ssd** directory  
with **datadir** subdirectory created within (user **eth** should be the owner).

#### switching to 64bit kernel

Because of memory allocation problem with the latest Raspbian and RPi4  
you should switch to 64bit kernel - by executing the **ethbian-64bit.sh** script.  
If you don't - when fully synced, geth will crash every couple of minutes/hours  
causing your Pi to hang eventually.

#### geth upgrade

The "**ethbian-geth-admin.sh -u**" will upgrade your geth to the latest version.

#### system/geth monitoring

Ethbian monitoring:

- collectd (collects system/geth data)
- geth_peers_geo2influx.py script (geolocation, eth's crontab)
- eth_price2influx.py script (eth price, eth's crontab)
- influx database (data storage)
- grafana (data visualization)

Use the **ethbian-monitoring.sh** command do disable/enable these services  
or check their status.

#### geth & RPi stats dashboard

When monitoring services are running you can monitor RPi temperature, system load  
and memory usage and some basic geth stats with a web browser:  
Grafana is running on port 3000 with default user **admin** and password **admin**.  
Just launch Firefox/Chrome/Safari and enter the following URL:  
**http<nolink>://pi's_IP_address:3000**, eg. http<nolink>://192.168.1.33:3000

#### start geth

With network configured and disk mounted you are ready to go.  
**sudo systemctl start geth** will start geth.  
**sudo systemctl enable geth** will make it start automatically on boot.

#### files

**/lib/systemd/system/geth.service** :geth settings  
**/usr/local/bin/geth/geth** :geth binary.  
**/usr/local/sbin** :admin scripts

**temp** :shows CPU temperature  
**gat** :attach to geth console  
**gsync** :shows geth synchronization progress  
**/var/log/geth.log** :geth logs

For more details visit [https://ethbian.org](https://ethbian.org)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## [v0.6] - 2020-02-16

### Changed

- Raspbian February 2020
- Geth 1.9.10
- ethbian_install.sh

## [v0.5] - 2019-12-27

### Added

- ethbian-64bit.sh

### Changed

- ethbian-ssd-init.sh

## [v0.4] - 2019-12-21

### Added

- logrotate configs
- eth_price2influx.py
- grafana clock panel plugin
- gsync script

### Changed

- collectd config file
- ethbian-monitoring.sh
- ethbian_install.sh
- ethbian_install.test.sh
- temp(erature) script

### Removed

- collectd df plugin

## [v0.3] - 2019-12-15

### Added

- ethbian-geth-admin.sh
- ethbian-monitoring.sh
- geth_peers_geo2influx.py
- grafana worldmap plugin

### Removed

- ethbian-geth-upgrade.sh

### Changed

- ethbian_install.sh
- ethbian_install.test.sh

## [v0.2] - 2019-12-05

### Added

- influxdb
- collectd
- grafana
- ethbian-geth-upgrade.sh

### Removed

- webfs

## [v0.1] - 2019-11-17

- initial release


ethbian_install.test.sh
#!/bin/bash

echo ""
echo "*****************************************"
echo "*       ETHBIAN INSTALL TEST v0.6       *"
echo "*****************************************"
echo ""
echo "*RR - requires reboot"
echo ""

# ------------------- functions ------------------
function file_exists () {
    echo "Checking if $2 exists"
    if [ $1 $3 ]; then
        echo '          [OK]'
    else
        echo '          [ERROR]'
    fi
}

function exec_code_check () {
    echo $2
    eval "$1"
    if [ $? -ne 0 ]; then
        echo '          [ERROR]'
    else
        echo '          [OK]'
    fi
}

function exec_output_check () {
    echo $3
    OUTPUT=$(eval "$2")
    if [ $OUTPUT -ne $1 ]; then
        echo '          [ERROR]'
    else
        echo '          [OK]'
    fi
}

# --------------------- basic ethbian checks ----------------
exec_code_check 'id eth > /dev/null 2>&1' 'Checking if user eth exists: '
file_exists '-x' 'temp script' '/usr/local/bin/temp'
file_exists '-x' 'gsync script' '/usr/local/bin/gsync'
file_exists '-x' 'gat script' '/usr/local/bin/gat'
exec_output_check 0 'dpkg -l| grep dphys-swapfile| wc -l' 'Checking if dphys-swapfile package was removed: '
exec_output_check 0 'swapon -s |wc -w' 'Checking if swap is disabled (*RR): '
file_exists '! -L' 'wpa_supplicant does not' '/etc/systemd/system/multi-user.target.wants/wpa_supplicant.service'
exec_output_check 1 "ulimit -a |grep 'open files' |grep -c 32000" 'Checking if open files limit is increased (*RR): '
exec_output_check 0 'ifconfig |grep -c inet6' 'Checking if IPv6 is disabled (*RR): '
exec_output_check 16 '/usr/bin/vcgencmd get_mem gpu |cut -f 2 -d "=" |sed "s/.$//"' 'Checking GPUs memory (*RR): '
echo ''

# --------------------- hardware ----------------
exec_output_check 0 'iw dev |grep -c Interface' 'Checking if wifi is disabled: '
exec_output_check 0 'hciconfig |wc -l' 'Checking if bluetooth is disabled: '
exec_output_check 0 'aplay -l 2>/dev/null|wc -l' 'Checking if sound card is disabled: '
exec_output_check 0 'lsmod |grep -c bcm2835' 'Checking if camera modules are disabled: '
echo ''

# --------------------- geth ----------------
file_exists '-x' 'geth binary' '/usr/local/bin/geth/geth'
file_exists '-f' 'geth.service file' '/lib/systemd/system/geth.service'
exec_output_check 2 'grep -c geth /etc/rsyslog.conf' 'Checking syslog for geth: '
exec_code_check 'grep geth /etc/logrotate.d/geth 1> /dev/null' 'Checking logrotate for geth: '
echo ''

# --------------------- sudo ---------------
exec_output_check 1 "sudo grep -c 'eth ALL=(ALL) NOPASSWD:/opt/vc/bin/vcgencmd measure_temp' /etc/sudoers" \
    'Checking sudo configuration: '
exec_output_check 1 "sudo grep -c 'eth ALL=(ALL) NOPASSWD:/usr/bin/tail /var/log/syslog' /etc/sudoers" ''
exec_output_check 1 "sudo grep -c 'eth ALL=(ALL) NOPASSWD:/usr/bin/tail /var/log/geth.log' /etc/sudoers" ''

# --------------------- admin scripts ---------------
file_exists '-x' 'ethbian-net.sh script' '/usr/local/sbin/ethbian-net.sh'
file_exists '-x' 'ethbian-ssd-init.sh script' '/usr/local/sbin/ethbian-ssd-init.sh'
file_exists '-x' 'ethbian-geth-admin.sh script' '/usr/local/sbin/ethbian-geth-admin.sh'
file_exists '-x' 'ethbian-monitoring.sh script' '/usr/local/sbin/ethbian-monitoring.sh'
file_exists '-x' 'ethbian-64bit.sh script' '/usr/local/sbin/ethbian-64bit.sh'
echo ''

# --------------------- influx -----------------
file_exists '-f' 'influxdb.conf file' '/etc/influxdb/influxdb.conf'
file_exists '-f' 'influxdb.conf.org file' '/etc/influxdb/influxdb.conf.org'
exec_code_check 'grep GOMAXPROCS /etc/default/influxdb 1> /dev/null' 'Checking GOMAXPROCS for influxdb: '
exec_code_check 'systemctl is-enabled --quiet influxdb' 'Checking if influxdb service is enabled: '
exec_code_check 'systemctl is-active --quiet influxdb' 'Checking if influxdb service is running: '
exec_output_check 2 'grep -c influxd /etc/rsyslog.conf' 'Checking syslog for influx: '
exec_code_check 'grep influx /etc/logrotate.d/influx 1> /dev/null' 'Checking logrotate for influx: '
echo ''

# --------------------- collectd ----------------
file_exists '-f' 'collectd.conf file' '/etc/collectd/collectd.conf'
file_exists '-f' 'collectd.conf.org file' '/etc/collectd/collectd.conf.org'
file_exists '-d' 'collectd directory' '/usr/local/lib/collectd'
file_exists '-f' 'rpi_temperature.py file' '/usr/local/lib/collectd/rpi_temperature.py'
file_exists '-f' 'geth_status.py file' '/usr/local/lib/collectd/geth_status.py'
exec_code_check 'systemctl is-enabled --quiet collectd' 'Checking if collectd service is enabled: '
exec_code_check 'systemctl is-active --quiet collectd' 'Checking if collectd service is running: '
exec_code_check 'grep collectd /etc/logrotate.d/collectd 1> /dev/null' 'Checking logrotate for collectd: '
echo ''

# --------------------- geth_peers_geo2influx ----------------
file_exists '-f' 'geolite_city.mmdb file' '/usr/local/lib/collectd/geolite_city.mmdb'
file_exists '-f' 'geth_peers_geo2influx.py file' '/usr/local/bin/geth_peers_geo2influx.py'
file_exists '-f' '/var/log/geo2influx.log file' '/var/log/geo2influx.log'
exec_output_check 1 'sudo grep -c geth_peers_geo2influx.py /var/spool/cron/crontabs/eth' 'Checking crontab for the eth user: '
exec_code_check 'grep geo2influx /etc/logrotate.d/geo2influx 1> /dev/null' 'Checking logrotate for geo2influx: '

# --------------------- eth_price2influx ----------------
file_exists '-f' 'eth_price2influx.py file' '/usr/local/bin/eth_price2influx.py'
file_exists '-f' '/var/log/price2influx.log file' '/var/log/price2influx.log'
exec_output_check 1 'sudo grep -c eth_price2influx.py /var/spool/cron/crontabs/eth' 'Checking crontab for the eth user: '
exec_code_check 'grep price2influx /etc/logrotate.d/price2influx 1> /dev/null' 'Checking logrotate for eth_price2influx: '

# --------------------- grafana ----------------
file_exists '-f' 'grafana.ini file' '/etc/grafana/grafana.ini'
file_exists '-f' 'grafana.ini.org file' '/etc/grafana/grafana.ini.org'
exec_output_check 1 'sudo grafana-cli plugins ls |grep -c "grafana-worldmap-panel"' 'Checking grafana worldmap plugin: '
exec_output_check 1 'sudo grafana-cli plugins ls |grep -c "grafana-clock-panel"' 'Checking grafana clock plugin: '
exec_output_check 1 'curl -s -X GET -u admin:admin "http://127.0.0.1:3000/api/datasources" |grep -c InfluxDB' \
    'Checking if grafana imported datasource: '
exec_output_check 1 'curl -s -X GET -u admin:admin "http://127.0.0.1:3000/api/search/" |grep -c -E "eth_price.*geth_peers.*geth_status"' \
    'Checking if grafana imported dashboards: '
exec_output_check 1 'curl -s -X GET -u admin:admin "http://127.0.0.1:3000/api/search?starred=true" |grep -c -E "geth_peers.*geth_status"' \
    'Checking if grafana starred dashboards: '
exec_code_check 'systemctl is-enabled --quiet grafana-server' 'Checking if grafana service is enabled: '
exec_code_check 'systemctl is-active --quiet grafana-server' 'Checking if grafana service is running: '


status2html-cron.sh
#!/bin/bash

# *****************************************
# *       ETHBIAN STATUS2HTML   v0.1      *
# *****************************************

TARGET=/var/www/html/index.html

NOW=`date +'%A, %T'`
UPTIME=`uptime -p`
DF_ROOT=`df -h /`
LOAD=`cat /proc/loadavg |cut -f -3 -d " "`
MEMORY=`free -m| grep -v Swap`

DF_SSD='NOT MOUNTED'
CMD=`mountpoint -q /mnt/ssd`
if [ $? -eq 0 ]; then
  DF_SSD=`df -h /mnt/ssd`
fi

TEMP=`sudo -n /opt/vc/bin/vcgencmd measure_temp`
if [ $? -ne 0 ]; then
  TEMP='NO DATA (or permissions)'
fi

GETH_STATUS=`systemctl --no-pager --output=cat status geth`

if [ -f /var/log/geth.log ]; then
  TAIL_GETH=`sudo /usr/bin/tail /var/log/geth.log |cut -f 1-3,8- -d " "|sed 's/ \+/ /g'`
else
  TAIL_GETH='NO DATA (or permissions)'
fi

TAIL_SYSLOG=`sudo -n /usr/bin/tail /var/log/syslog`
if [ $? -ne 0 ]; then
  TAIL_SYSLOG='NO DATA (or permissions)'
fi

GETH_SYNCING=""
GETH_NODEINFO=""
CMD=`pgrep geth`
if [ $? -eq 0 ]; then
  GETH_SYNCING=`/usr/local/bin/geth/geth attach --datadir=/mnt/ssd/datadir --exec eth.syncing |grep -Ev '{|}'`
  GETH_NODEINFO=`/usr/local/bin/geth/geth attach --datadir=/mnt/ssd/datadir --exec admin.nodeInfo |grep -Ev 'enode:|enr:|{|}'`
fi
if [ "$GETH_SYNCING" = "" ]; then
  GETH_SYNCING="NO DATA"
fi
if [ "$GETH_NODEINFO" = "" ]; then
  GETH_NODEINFO="NO DATA"
fi

echo '<!DOCTYPE html><head><title>Ethbian status</title><link rel="stylesheet" type="text/css" href="main.css" />' > $TARGET
echo '<meta charset="utf-8"></head><body><table>' >> $TARGET
echo "<tr><th>generated:</th><td>$NOW</td></tr>" >> $TARGET
echo "<tr><td colspan=\"2\"></td></tr>" >> $TARGET

echo "<tr><th>server uptime:</th><td>$UPTIME</td></tr>" >> $TARGET
echo "<tr><th>system load:</th><td>$LOAD</td></tr>" >> $TARGET
echo "<tr><th>temperature:</th><td>$TEMP</td></tr>" >> $TARGET
echo "<tr><th>root FS:</th><td><pre>$DF_ROOT</pre></td></tr>" >> $TARGET
echo "<tr><th>eth data FS:</th><td><pre>$DF_SSD</pre></td></tr>" >> $TARGET
echo "<tr><th>RAM (in MiB):</th><td><pre>$MEMORY</pre></td></tr>" >> $TARGET

echo "<tr><td colspan=\"2\"></td></tr>" >> $TARGET
echo "<tr><th>geth<br/>status</th><td><pre>$GETH_STATUS</pre></td></tr>" >> $TARGET

echo "<tr><td colspan=\"2\"></td></tr>" >> $TARGET
echo "<tr><th>geth<br/>syncing</th><td><pre>$GETH_SYNCING</pre></td></tr>" >> $TARGET

echo "<tr><td colspan=\"2\"></td></tr>" >> $TARGET
echo "<tr><th>geth.log<br/>(last 10 lines)</th><td><pre>$TAIL_GETH</pre></td></tr>" >> $TARGET

echo "<tr><td colspan=\"2\"></td></tr>" >> $TARGET
echo "<tr><th>syslog<br/>(last 10 lines)</th><td><pre>$TAIL_SYSLOG</pre></td></tr>" >> $TARGET

echo "<tr><td colspan=\"2\"></td></tr>" >> $TARGET
echo "<tr><th>geth<br/>nodeInfo</th><td><pre>$GETH_NODEINFO</pre></td></tr>" >> $TARGET

echo "</table></body></html>" >> $TARGET


admin/scripts/ethbian-64bit.sh
#!/bin/bash

echo ""
echo "*****************************************"
echo "*          ETHBIAN 64BIT v0.1           *"
echo "*****************************************"
echo ""

CURRENT=`uname -m`
if [ "$CURRENT" = "aarch64" ]; then
  echo ""
  echo "You're already running 64bit kernel."
  echo ""
  exit 1
fi

grep -q ^'arm_64bit=1' /boot/config.txt
if [ $? -eq 0 ]; then
  echo ""
  echo "Your system is ready."
  echo "Please restart the box."
  echo ""
  exit 1
fi

echo "### updating the package list..."
echo ""
sudo apt-get -y update
echo ""
echo "### full-upgrading the system..."
echo ""
sudo apt-get -y full-upgrade
echo ""
echo "### updating the bootloader EEPROM"
echo ""
sudo rpi-eeprom-update -a
echo ""
echo "### switching to 64bit kernel"
echo ""
sudo /bin/bash -c 'echo arm_64bit=1 >> /boot/config.txt'
echo ""
echo "Done. Please restart the box."
echo ""

ethbian_install.sh
#!/bin/bash

echo ""
echo "*****************************************"
echo "*    ETHBIAN SD CARD IMAGE SETUP v0.6   *"
echo "*****************************************"
echo ""

function create_logrotate_config () {
  sudo /bin/bash -c "cat << EOF > /etc/logrotate.d/$1
/var/log/$2
{
  rotate 7
  daily
  missingok
  notifempty
  delaycompress
  compress
  postrotate
    /usr/lib/rsyslog/rsyslog-rotate
  endscript
}
EOF"
}

echo -n "### Detecting CPU architecture... "
isARM=$(uname -m | grep -c 'arm')
if [ ${isARM} -eq 0 ]; then
  echo "!!! FAIL !!!"
  echo "   This version can only run on ARM architecture"
  exit 1
else
  echo "OK"
fi

echo -n "### Detecting base image... "
isRaspbian=$(cat /etc/os-release 2>/dev/null | grep -c 'Raspbian')
if [ ${isRaspbian} -eq 0 ]; then
  echo "!!! FAIL !!!"
  echo "   This version can only run on Raspbian"
  exit 1
else
  echo "OK"
fi

echo ""
echo "### Updating system..."
echo ""
sudo apt-get -y update
echo ""
echo "### Upgrading system..."
echo "   this may take a while"
echo ""
sudo apt-get -y upgrade
echo ""
echo "### Update & upgrade finished"
echo ""
echo ""


echo "### Config"

echo "  # adding eth user..."
sudo adduser --shell /bin/bash --gecos "" --disabled-login eth
echo ""

echo "  # changing hostname..."
sudo /bin/bash -c 'echo "ethbian" > /etc/hostname'
sudo sed -i "s/^127.0.1.1.*/127.0.1.1\tethbian/" /etc/hosts
echo ""

echo "  # changing passwords..."
echo "pi:ethbian" | sudo chpasswd
sudo passwd -e pi
echo ""

echo "  # installing tools..."
echo ""
sudo apt-get install -y git jq dstat lsof nmap screen tmux fail2ban dialog sysstat ipcalc sqlite3 software-properties-common python-requests
if [ ! -d /mnt/ssd ]; then
  sudo mkdir /mnt/ssd
fi

sudo /bin/bash -c 'cat << EOF > /usr/local/bin/gsync
#!/bin/sh
echo 'eth.syncing' |sudo gat |grep tBlock
EOF'
sudo chmod +x /usr/local/bin/gsync

/bin/bash -c 'cat << EOF >> /home/pi/.bashrc

# show pi temperature
echo 
echo ----------------------------------
/usr/local/bin/temp
echo ----------------------------------
EOF'

sudo /bin/bash -c 'cat << EOF > /usr/local/bin/gat
#!/bin/sh
/usr/local/bin/geth/geth attach --datadir=/mnt/ssd/datadir
EOF'
sudo chmod +x /usr/local/bin/gat

sudo /bin/bash -c 'cat << EOF > /etc/motd

    --- Welcome to Ethbian! ---
               v0.6

admin commands (for the 'pi' user):
  ethbian-64bit.sh - switch to 64bit kernel
  ethbian-net.sh - simple network configuration
  ethbian-geth-admin.sh - upgrade geth binary
  ethbian-ssd-init.sh - ssd drive init
  ethbian-monitoring.sh - system/geth monitoring on/off

after switching to 64bit kernel, configuring network and ssd drive:
- to start geth: sudo systemctl start geth
- to run geth on startup: sudo systemctl enable geth

grafana is running on port 3000
(user: admin, password: admin)

SSD drive is a must.
Active cooling is highly recommended.

For more details visit https://ethbian.org

EOF'

echo ""
echo -e "\nalias gat='sudo /usr/local/bin/gat'" >> /home/pi/.bashrc

cd /tmp
git clone https://github.com/ethbian/ethbian.git
cd ethbian

chmod +x admin/scripts/*
sudo chown root:root admin/scripts/*
sudo mv admin/scripts/temp /usr/local/bin
sudo mv admin/scripts/* /usr/local/sbin

echo "  # disabling swap..."
echo ""
sudo dphys-swapfile swapoff
sudo dphys-swapfile uninstall
sudo apt-get purge -y dphys-swapfile
echo ""

echo "  # removing wpa_supplicant..."
echo ""
sudo systemctl disable wpa_supplicant
echo ""

echo "   # increasing open files limits..."
sudo /bin/bash -c 'cat << EOF > /etc/security/limits.d/90-geth.conf
*    soft nofile 32000
*    hard nofile 32000
root soft nofile 32000
root hard nofile 32000
EOF'
echo ""

echo "  # disabling IPv6..."
sudo /bin/bash -c 'cat << EOF >> /etc/sysctl.d/99-sysctl.conf
# disable ipv6
net.ipv6.conf.all.disable_ipv6 = 1
net.ipv6.conf.default.disable_ipv6 = 1
net.ipv6.conf.lo.disable_ipv6 = 1
EOF'
echo ""

echo "### Hardware"

echo "  # disabling wifi..."
sudo /bin/bash -c 'echo "dtoverlay=disable-wifi" >> /boot/config.txt'
echo ""

echo "  # disabling bluetooth..."
sudo /bin/bash -c 'echo "dtoverlay=disable-bt" >> /boot/config.txt'
echo ""

echo "  # disabling sound card..."
sudo sed -i 's/dtparam=audio=on/dtparam=audio=off/' /boot/config.txt
echo ""

echo "  # reassigning memory from GPU to CPU..."
sudo /bin/bash -c 'echo "gpu_mem=16" >> /boot/config.txt'
echo ""

echo "  # disabling camera modules..."
sudo /bin/bash -c 'echo -e "blacklist bcm2835_codec\nblacklist bcm2835_v4l2" > /etc/modprobe.d/disable_rpi4_camera.conf'
echo ""

sudo /bin/bash -c "echo 'eth ALL=(ALL) NOPASSWD:/opt/vc/bin/vcgencmd measure_temp' >> /etc/sudoers"
sudo /bin/bash -c "echo 'eth ALL=(ALL) NOPASSWD:/usr/bin/tail /var/log/syslog' >> /etc/sudoers"
sudo /bin/bash -c "echo 'eth ALL=(ALL) NOPASSWD:/usr/bin/tail /var/log/geth.log' >> /etc/sudoers"

sudo sed -i '/^exit/itest -f /etc/ssh/ssh_host_dsa_key || dpkg-reconfigure openssh-server' /etc/rc.local

echo "### GETH"
/usr/local/sbin/ethbian-geth-admin.sh -i

echo "  # syslog..."
sudo sed -i "/^auth/i :programname, isequal, \"geth\" \/var\/log\/geth.log" /etc/rsyslog.conf
sudo sed -i "/^auth/i :programname, isequal, \"geth\" stop" /etc/rsyslog.conf

create_logrotate_config 'geth' 'geth.log'

echo "### Monitoring"
GITHUB_RPI_TEMP='https://raw.githubusercontent.com/ethbian/rpi_temperature_plugin4collectd/master/rpi_temperature.py'
GITHUB_GETH_STATUS='https://raw.githubusercontent.com/ethbian/geth_status_plugin4collectd/master/geth_status.py'

echo ""
echo "  # installing monitoring tools..."
sudo apt-get install -y collectd collectd-utils influxdb influxdb-client python-influxdb python-geoip2
cd /tmp/ethbian
echo ""

echo "  # influx..."
sudo systemctl stop influxdb
sudo mv /etc/influxdb/influxdb.conf /etc/influxdb/influxdb.conf.org
sudo mv admin/conf/influxdb.conf /etc/influxdb/
sudo /bin/bash -c 'echo "GOMAXPROCS=1" >> /etc/default/influxdb'
sudo systemctl enable influxdb
sudo sed -i "/^auth/i :programname, isequal, \"influxd\" \/var\/log\/influx.log" /etc/rsyslog.conf
sudo sed -i "/^auth/i :programname, isequal, \"influxd\" stop" /etc/rsyslog.conf
create_logrotate_config 'influx' 'influx.log'
echo ""

echo "  # collectd..."
sudo systemctl stop collectd
sudo mv /etc/collectd/collectd.conf /etc/collectd/collectd.conf.org
sudo mv admin/conf/collectd.conf /etc/collectd/
sudo mkdir /usr/local/lib/collectd
cd /usr/local/lib/collectd
sudo wget $GITHUB_RPI_TEMP
sudo wget $GITHUB_GETH_STATUS
sudo systemctl enable collectd
create_logrotate_config 'collectd' 'collectd.log'
echo ""

echo "  # geth_peers_geo2influx..."
GEO_SCRIPT='geth_peers_geo2influx.py'
GEO_LOG='/var/log/geo2influx.log'
GITHUB_GEO2INFLUX='https://raw.githubusercontent.com/ethbian/geth_peers_geo2influx/master/'
sudo touch $GEO_LOG
sudo chown eth $GEO_LOG
create_logrotate_config 'geo2influx' 'geo2influx.log'
cd /usr/local/bin
sudo wget ${GITHUB_GEO2INFLUX}${GEO_SCRIPT}
sudo chmod +x $GEO_SCRIPT

sudo touch /var/spool/cron/crontabs/eth
sudo chown eth:crontab /var/spool/cron/crontabs/eth
sudo chmod 0600 /var/spool/cron/crontabs/eth
sudo /bin/bash -c "echo 'SHELL=/bin/bash' > /var/spool/cron/crontabs/eth"
sudo /bin/bash -c "echo '*/30 * * * * /usr/local/bin/$GEO_SCRIPT >> $GEO_LOG 2>&1' >> /var/spool/cron/crontabs/eth"

echo "  # eth_price2influx..."
PRICE_LOG='/var/log/price2influx.log'
PRICE_SCRIPT='eth_price2influx.py'
GITHUB_PRICE2INFLUX='https://raw.githubusercontent.com/ethbian/eth_price2influx/master/eth_price2influx.py'

sudo touch $PRICE_LOG
sudo chown eth $PRICE_LOG
create_logrotate_config 'price2influx' 'price2influx.log'
cd /usr/local/bin
sudo wget $GITHUB_PRICE2INFLUX
sudo chmod +x $PRICE_SCRIPT
sudo /bin/bash -c "echo '*/5 * * * * /usr/local/bin/$PRICE_SCRIPT >> $PRICE_LOG 2>&1' >> /var/spool/cron/crontabs/eth"

echo "  # grafana..."
cd /tmp/ethbian
wget -q -O - https://packages.grafana.com/gpg.key | sudo apt-key add -
sudo /bin/bash -c 'echo "deb https://packages.grafana.com/oss/deb stable main" > /etc/apt/sources.list.d/grafana.list'
sudo apt-get update
sudo apt-get install -y grafana
sudo systemctl stop grafana-server
sleep 10

sudo cp /etc/grafana/grafana.ini /etc/grafana/grafana.ini.org
sudo sed -i 's/^;reporting_enabled = true/reporting_enabled = false/' /etc/grafana/grafana.ini
sudo sed -i 's/^;check_for_updates = true/check_for_updates = false/' /etc/grafana/grafana.ini
sudo sed -i 's/^;level = info/level = warn/' /etc/grafana/grafana.ini
sudo sed -i '/Enable internal metrics/!b;n;cenabled = false' /etc/grafana/grafana.ini
sudo sed -i 's/^;disable_total_stats = false/disable_total_stats = true/' /etc/grafana/grafana.ini
sudo sed -i 's/^;allow_sign_up = true/allow_sign_up = false/' /etc/grafana/grafana.ini
sudo systemctl daemon-reload
sudo systemctl enable grafana-server
sudo systemctl start grafana-server
sleep 10
sudo systemctl stop grafana-server
sleep 5
sudo grafana-cli plugins install grafana-worldmap-panel
sudo grafana-cli plugins install grafana-clock-panel
cat admin/conf/grafana_ds_influx.sql | sudo sqlite3 /var/lib/grafana/grafana.db
cat admin/conf/grafana_dash_geth_status.sql | sudo sqlite3 /var/lib/grafana/grafana.db
cat admin/conf/grafana_dash_geth_peers.sql | sudo sqlite3 /var/lib/grafana/grafana.db
cat admin/conf/grafana_dash_eth_price.sql | sudo sqlite3 /var/lib/grafana/grafana.db
cat admin/conf/grafana_star.sql | sudo sqlite3 /var/lib/grafana/grafana.db

echo "### Cleaning up"
sudo apt-get purge -y avahi-daemon mariadb-common mysql-common libvirt0 openjdk-11-jre-headless adwaita-icon-theme
sudo apt -y autoremove
sudo apt clean

echo "### Done."
